spdlog/src/formatter/json_formatter.rs

```diff
@@ -0,0 +1,190 @@
+use std::{
+    fmt::{self, Write},
+    marker::PhantomData,
+    time::SystemTime,
+};
+
+use cfg_if::cfg_if;
+use serde::{ser::SerializeStruct, Serialize};
+
+use crate::{
+    formatter::{FmtExtraInfo, Formatter},
+    Error, Record, StringBuf, __EOL,
+};
+
+struct JsonRecord<'a>(&'a Record<'a>);
+
+impl<'a> Serialize for JsonRecord<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let src_loc = self.0.source_location();
+
+        let mut record =
+            serializer.serialize_struct("JsonRecord", if src_loc.is_none() { 4 } else { 5 })?;
+
+        record.serialize_field("level", &self.0.level())?;
+        record.serialize_field(
+            "timestamp",
+            &self
+                .0
+                .time()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .expect("invalid timestamp")
+                .as_secs()
+                .to_string(),
+        )?;
```

Comment:
```suggestion
        record.serialize_field(
            "timestamp",
            &self
                .0
                .time()
                .duration_since(SystemTime::UNIX_EPOCH)
                .ok()
                .and_then(|dur| u64::try_from(dur.as_millis()).ok())
                .expect("invalid timestamp"),
        )?;
```

I've discussed with @NotEvenANeko in private messages.

Initially I simply thought that UNIX timestamp would be OK, but then realized that it is in seconds, which is obviously not precise enough for logging. After some investigation, I think a millisecond timestamp might be more appropriate, based on the following clues:

- in JavaScript, [`Date::now`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now) and [the constructor of `Date`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#value) returns and accepts *a number of milliseconds since January 1, 1970 00:00:00 UTC, with leap seconds ignored*.

- Most programming languages have time libraries that support parsing millisecond timestamps.

- Nanoseconds are more precise, but it uses more significant integer digits, and perhaps this precision is unnecessary.

Since this integer will eventually be expressed in JSON, and [`Duration::as_millis`](https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.as_millis) returns a `u128`, we need to convert it to `u64` to express it in JSON. Although the maximum safe type for JSON integers in JavaScript is `i53`, I tested it with Rust `chrono` crate, and the maximum millisecond timestamp it supports parsing doesn't even exceed `i53::MAX`, and that maximum timestamp has reached the year 262142. So in my opinion, representing it as a string is not necessary for the foreseeable future.

```rust
extern crate chrono; // 0.4.38

use chrono::prelude::*;

fn main() {
    const JS_MAX_SAFE_INTEGER: i64 = 9007199254740991;
    let max_utc = DateTime::<Utc>::MAX_UTC;
    let max_timestamp = max_utc.timestamp_millis();
    println!("Max UTC supported by chrono   : {}", max_utc);
    println!("Max timestamp in millisecond  : {}", max_timestamp);
    println!("Max safe integer in JavaScript: {}", JS_MAX_SAFE_INTEGER);
    println!("Max u64                       : {}", u64::MAX);
    assert!(max_timestamp <= JS_MAX_SAFE_INTEGER);
}
```

```
Max UTC supported by chrono   : +262142-12-31 23:59:59.999999999 UTC
Max timestamp in millisecond  : 8210266876799999
Max safe integer in JavaScript: 9007199254740991
Max u64                       : 18446744073709551615
```

Using `u64::try_from` here so we can catch the overflow error instead of  having it wrapped implicitly.